overlord/patch/patch.go

```diff
@@ -40,45 +49,147 @@ func Init(s *state.State) {
+		return err
+	}
+	if len(snapst.Sequence) < 2 {
+		return nil
+	}
+	currentIndex := snapst.LastIndex(snapst.Current)
+	if currentIndex < 0 {
+		return fmt.Errorf("internal error: couldn't find current core revision in the snap sequence")
+	}
+
+	prevRev := snapst.Sequence[currentIndex-1].Revision
+
+	// check if core revision is for patch level 6.0.
+	// the revision number here is the highest rev number from candidate channels of all architectures
+	if prevRev.N > 5332 {
+		return nil
+	}
+
+	var sublevelResetTime time.Time
+	var lastRefresh time.Time
+	if err := s.Get("last-refresh", &lastRefresh); err != nil && err != state.ErrNoState {
+		return fmt.Errorf("cannot read last-refresh: %s", err)
+	}
+	err = s.Get("patch-sublevel-reset", &sublevelResetTime)
+	if err != nil && err != state.ErrNoState {
+		return fmt.Errorf("cannot read patch-sublevel-reset: %s", err)
+	}
+	if sublevelResetTime != lastRefresh || err == state.ErrNoState {
+		*sublevel = 0
+		s.Set("patch-sublevel", *sublevel)
+		s.Set("patch-sublevel-reset", lastRefresh)
+	}
+
+	return nil
 }
 
 // Apply applies any necessary patches to update the provided state to
 // conventions required by the current patch level of the system.
 func Apply(s *state.State) error {
-	var stateLevel int
+	var stateLevel, stateSublevel int
 	s.Lock()
 	err := s.Get("patch-level", &stateLevel)
+	if err == nil || err == state.ErrNoState {
+		err = s.Get("patch-sublevel", &stateSublevel)
+	}
 	s.Unlock()
+
 	if err != nil && err != state.ErrNoState {
 		return err
 	}
-	if stateLevel == Level {
-		// already at right level, nothing to do
-		return nil
-	}
+
 	if stateLevel > Level {
 		return fmt.Errorf("cannot downgrade: snapd is too old for the current system state (patch level %d)", stateLevel)
 	}
 
-	level := stateLevel
-	for level < Level {
+	// check we refreshed from 6.0 which was not aware of sublevels
+	if stateLevel == 6 && stateSublevel > 0 {
+		if err := maybeResetSublevelForLevel60(s, &stateSublevel); err != nil {
+			return err
+		}
+	}
+
+	if stateLevel == Level && stateSublevel == Sublevel {
+		return nil
+	}
+
+	// downgrade within same level; update sublevel in the state so that sublevel patches
+	// are re-applied if the user refreshes to a newer patch sublevel again.
+	if stateLevel == Level && stateSublevel > Sublevel {
+		s.Lock()
+		s.Set("patch-sublevel", Sublevel)
+		s.Unlock()
+		return nil
+	}
+
+	// apply any missing sublevel patches for current state level before upgrading to new levels.
+	// the 0th sublevel patch is a patch for major level update (e.g. 7.0),
+	// therefore there is +1 for the indices.
+	if stateSublevel+1 < len(patches[stateLevel]) {
+		if err := applySublevelPatches(stateLevel, stateSublevel+1, s); err != nil {
+			return err
+		}
+	}
+
+	// at the lower Level - apply all new level and sublevel patches
+	for level := stateLevel; level < Level; level++ {
+		sublevels := patches[level+1]
```

Comment:
the `+1` could be moved to the for loop
  
    level := stateLevel+1 ; level <= Level

no?

Reply:
Yes. Done (small downside of doing it is "-1" needed for the debug messages though).